From 338c2b90dff6f57a197b12f1d0df11f7d842f5bd Mon Sep 17 00:00:00 2001
From: Lukas Nykryn <lnykryn@redhat.com>
Date: Tue, 12 Feb 2019 16:58:16 +0100
Subject: [PATCH] rules: add elevator= kernel command line parameter

Kernel removed the elevator= option, so let's reintroduce
it for rhel8 via udev rule.

RHEL-only: feature

Related: RHEL-40924
---
 rules.d/40-elevator.rules | 20 ++++++++++++++++++++
 rules.d/meson.build       |  1 +
 2 files changed, 21 insertions(+)
 create mode 100644 rules.d/40-elevator.rules

diff --git a/rules.d/40-elevator.rules b/rules.d/40-elevator.rules
new file mode 100644
index 0000000000..dbe8fc81a4
--- /dev/null
+++ b/rules.d/40-elevator.rules
@@ -0,0 +1,20 @@
+# We aren't adding devices skip the elevator check
+ACTION!="add", GOTO="sched_out"
+
+SUBSYSTEM!="block", GOTO="sched_out"
+ENV{DEVTYPE}!="disk", GOTO="sched_out"
+
+# Technically, dm-multipath can be configured to use an I/O scheduler.
+# However, there are races between the 'add' uevent and the linking in
+# of the queue/scheduler sysfs file.  For now, just skip dm- devices.
+KERNEL=="dm-*|md*", GOTO="sched_out"
+
+# Skip bio-based devices, which don't support an I/O scheduler.
+ATTR{queue/scheduler}=="none", GOTO="sched_out"
+
+# If elevator= is specified on the kernel command line, change the
+# scheduler to the one specified.
+IMPORT{cmdline}="elevator"
+ENV{elevator}!="", ATTR{queue/scheduler}="$env{elevator}"
+
+LABEL="sched_out"
\ No newline at end of file
diff --git a/rules.d/meson.build b/rules.d/meson.build
index edf419f449..2b57322a12 100644
--- a/rules.d/meson.build
+++ b/rules.d/meson.build
@@ -8,6 +8,7 @@ rules = [
         [files('40-redhat-hotplug.rules',
                '40-redhat-s390.rules',
                '40-redhat.rules',
+               '40-elevator.rules',
                '60-autosuspend.rules',
                '60-block.rules',
                '60-cdrom_id.rules',
